src/execution/operator/aggregate/physical_window.cpp

```diff
@@ -376,30 +421,196 @@ void WindowLocalSourceState::GeneratePartition(WindowGlobalSinkState &gstate, co
+
+	vector<LogicalType> output_types;
+	for (idx_t expr_idx = 0; expr_idx < op.select_list.size(); ++expr_idx) {
+		D_ASSERT(op.select_list[expr_idx]->GetExpressionClass() == ExpressionClass::BOUND_WINDOW);
+		auto &wexpr = op.select_list[expr_idx]->Cast<BoundWindowExpression>();
+		output_types.emplace_back(wexpr.return_type);
+	}
+	output_chunk.Initialize(Allocator::Get(gsource.context), output_types);
+}
+
+WindowGlobalSourceState::Task WindowGlobalSourceState::CreateTask(idx_t hash_bin) {
+	//	Build outside the lock so no one tries to steal before we are done.
+	auto partition_source = make_uniq<WindowPartitionSourceState>(context, *this);
+	partition_source->BuildPartition(gsink, hash_bin);
+	Task result(partition_source.get(), partition_source->GetScanner());
+
+	//	Is there any data to scan?
+	if (result.second) {
+		lock_guard<mutex> built_guard(built_lock);
+		built[hash_bin] = std::move(partition_source);
+
+		return result;
+	}
+
+	return Task();
+}
+
+WindowGlobalSourceState::Task WindowGlobalSourceState::StealWork() {
+	for (idx_t hash_bin = 0; hash_bin < built.size(); ++hash_bin) {
+		lock_guard<mutex> built_guard(built_lock);
+		auto &partition_source = built[hash_bin];
+		if (!partition_source) {
+			continue;
+		}
+
+		Task result(partition_source.get(), partition_source->GetScanner());
+
+		//	Is there any data to scan?
+		if (result.second) {
+			return result;
+		}
+	}
+
+	//	Nothing to steal
+	return Task();
+}
+
+WindowGlobalSourceState::Task WindowGlobalSourceState::NextTask(idx_t hash_bin) {
+	auto &hash_groups = gsink.global_partition->hash_groups;
+	const auto bin_count = built.size();
+
+	//	Flush unneeded data
+	if (hash_bin < bin_count) {
+		//	Lock and delete when all blocks have been scanned
+		//	We do this here instead of in NextScan so the WindowLocalSourceState
+		//	has a chance to delete its state objects first,
+		//	which may reference the partition_source
+
+		//	Delete data outside the lock in case it is slow
+		HashGroupSourcePtr killed;
+		lock_guard<mutex> built_guard(built_lock);
+		auto &partition_source = built[hash_bin];
+		if (partition_source && !partition_source->unscanned) {
+			killed = std::move(partition_source);
+		}
+	}
+
+	hash_bin = next_build++;
+	if (hash_bin < bin_count) {
+		//	Find a non-empty hash group.
+		for (; hash_bin < hash_groups.size(); hash_bin = next_build++) {
+			if (hash_groups[hash_bin]) {
+				auto result = CreateTask(hash_bin);
+				if (result.second) {
+					return result;
+				}
+			}
+		}
+
+		//	OVER() doesn't have a hash_group
+		if (hash_groups.empty()) {
+			auto result = CreateTask(hash_bin);
+			if (result.second) {
+				return result;
+			}
+		}
+	}
+
+	//	Work stealing
+	while (tasks_remaining) {
+		auto result = StealWork();
+		if (result.second) {
+			return result;
+		}
+
+		//	If there is nothing to steal but there are unfinished partitions,
+		//	yield until any pending builds are done.
+		std::this_thread::yield();
```

Comment:
nit: could we perhaps move this to the `TaskScheduler`, so this can be `TaskScheduler::Yield()` and we don't need to include `thread` here